core/shared/src/main/scala/zio/ZSchedule.scala

```diff
@@ -378,82 +395,108 @@ trait ZSchedule[-R, -A, +B] extends Serializable { self =>
+      val update  = self.update
     }
 
   /**
    * Returns a new schedule with the specified effectful modification
-   * applied to each delay produced by this schedule.
+   * applied to each sleep performed by this schedule.
+   *
+   * Note that this does not apply to sleeps performed in ZSchedule#initial.
+   * All effects executed while calculating the modified duration will run with the old
+   * environment.
    */
-  final def modifyDelay[R1 <: R](f: (B, Duration) => ZIO[R1, Nothing, Duration]): ZSchedule[R1, A, B] =
-    updated(
-      update =>
-        (a, s) =>
-          update(a, s).flatMap { step =>
-            f(step.finish(), step.delay).map(d => step.delayed(_ => d))
-          }
-    )
+  final def modifyDelay(
+    f: (B, Duration) => ZIO[ZEnv, Nothing, Duration]
+  )(
+    implicit ev: ZEnv <:< R
+  ): ZSchedule[ZEnv, A, B] = modifyDelay_(f)(g => ZEnv.mapClock(g) andThen ev.apply)
+
+  /**
+   * Returns a new schedule with the specified effectful modification
+   * applied to each sleep performed by this schedule.
+   *
+   * This version supports arbitrary environments.
+   *
+   * Note that this does not apply to sleeps performed in ZSchedule#initial.
+   * All effects executed while calculating the modified duration will run with the old
+   * environment.
+   */
+  final def modifyDelay_[R1](
+    f: (B, Duration) => ZIO[R1, Nothing, Duration]
+  )(
+    g: (Clock.Service[Any] => Clock.Service[Any]) => R1 => R
+  ): ZSchedule[R1 with Clock, A, B] = {
+    def proxy(clock0: Clock.Service[Any], env: R1, current: B): Clock.Service[Any] = new Clock.Service[Any] {
+      def currentTime(unit: TimeUnit) = clock0.currentTime(unit)
+      def currentDateTime             = clock0.currentDateTime
+      val nanoTime                    = clock0.nanoTime
+      def sleep(duration: Duration)   = f(current, duration).provide(env).flatMap(clock0.sleep)
+    }
+    new ZSchedule[R1 with Clock, A, B] {
+      type State = self.State
+      val initial = self.initial.provideSome[R1](g(identity))
+      val extract = (a: A, s: self.State) => self.extract(a, s)
+      val update = (a: A, s: self.State) =>
+        self.update(a, s).provideSome[R1](env => g(proxy(_, env, self.extract(a, s)))(env))
+    }
+  }
 
   /**
    * A new schedule that applies the current one but runs the specified effect
    * for every decision of this schedule. This can be used to create schedules
    * that log failures, decisions, or computed values.
    */
-  final def onDecision[A1 <: A](f: (A1, ZSchedule.Decision[State, B]) => UIO[Unit]): ZSchedule[R, A1, B] =
+  final def onDecision[A1 <: A](f: (A1, self.State) => UIO[Unit]): ZSchedule[R, A1, B] =
     updated(update => (a, s) => update(a, s).tap(step => f(a, step)))
 
   /**
-   * Returns a new schedule that reconsiders the decision made by this schedule.
+   * Provide all requirements to the schedule.
    */
-  final def reconsider[A1 <: A, C](
```

Comment:
It would be a shame to lose `reconsider` / `reconsiderM`, especially since we can keep them with `R = ZEnv`.